### PR TITLE
[Consensus] Snapshot uses QC storage

### PR DESCRIPTION
**Pull request by durkmurder**
https://github.com/dapperlabs/flow-go/issues/6484

### Context

This PR replaces implementation of `RandomSource` and `QuorumCertificate` to use QC storage instead of querying QCs from child block. As part of this PR removed separate storage for root QC, used instead newly implemented QC storage.

### CODE REVIEW

File: state/protocol/badger/snapshot.go [hunk 1]
```diff
@@ -47,80 +45,16 @@ func (s *Snapshot) Head() (*flow.Header, error) {
...
 	return qc, nil
 }
 
-// validChild returns a child of the snapshot head. Any valid child may be returned.
-// Subsequent calls are not guaranteed to return the same child.
-// Since blocks are fully validated before insertion to the state, all stored child
-// blocks are valid and may be returned.
-//
-// Error returns:
-//   - state.NoChildBlockError if no valid child exists.
```
Comment (jordanschalm) [inline state/protocol/badger/snapshot.go:106]: `NoChildBlockError` is no longer used - we can remove it

https://github.com/onflow/flow-go/blob/5c4e877dc3bdca7d8617bfee7f25c7c95e9660ae/state/errors.go#L93-L116

**Review by jordanschalm — APPROVED** [review]

File: state/protocol/badger/snapshot.go [hunk 2]
```diff
@@ -416,41 +350,19 @@ func (s *Snapshot) descendants(blockID flow.Identifier) ([]flow.Identifier, erro
...
-		}
-		return seed, nil
-	}
-
-	// CASE 2: for any other block, use any valid child
-	child, err := s.validChild()
-	if err != nil {
-		return nil, fmt.Errorf("failed to get valid child of block %x: %w", s.blockID, err)
+		return nil, err
 	}
```
Comment (AlexHentschel) [inline state/protocol/badger/snapshot.go:359]: I think this code is trivial enough that we can cut some white spaces:
```suggestion
```

File: state/protocol/badger/snapshot.go [hunk 3]
```diff
@@ -416,41 +350,19 @@ func (s *Snapshot) descendants(blockID flow.Identifier) ([]flow.Identifier, erro
...
 
-	seed, err := seed.FromParentQCSignature(child.ParentVoterSigData)
+	randomSource, err := seed.FromParentQCSignature(qc.SigData)
 	if err != nil {
-		return nil, fmt.Errorf("could not create seed from header's signature: %w", err)
+		return nil, fmt.Errorf("could not create seed from QC's signature: %w", err)
 	}
```
Comment (AlexHentschel) [inline state/protocol/badger/snapshot.go:364]:
```suggestion
```

File: state/protocol/badger/snapshot_test.go [hunk 4]
```diff
@@ -853,35 +854,10 @@ func TestQuorumCertificate(t *testing.T) {
 			require.Nil(t, err)
 
 			_, err = state.AtBlockID(block1.ID()).QuorumCertificate()
-			assert.Error(t, err)
+			assert.ErrorIs(t, err, storage.ErrNotFound)
```
Comment (AlexHentschel) [inline state/protocol/badger/snapshot_test.go:857]: ✨ thanks for testing for specific sentinel 🙇

File: state/protocol/snapshot.go [hunk 5]
```diff
@@ -102,11 +104,9 @@ type Snapshot interface {
 
 	// RandomSource returns the source of randomness derived from the Head block.
```
Comment (AlexHentschel) [inline state/protocol/snapshot.go:105]: I think this is a very subtle topic that we need to carefully explain.
```suggestion
	// RandomSource returns the source of randomness _for_ the snapshot's Head block. 
	// Note that the source of randomness for a block `H`, is contained in the 
	// QuorumCertificate [QC] for block `H` (QCs for H are distributed as part of child
	// blocks, timeout messages or timeout certificates). While there might be different 
	// QCs for block H, they all yield exactly the same source of randomness (feature of 
	// threshold signatures used here). Therefore, it is a possibility that there is no 
	// QC known (yet) for the head block. 
```

**Review by AlexHentschel — APPROVED** [review]
🏄‍♂️ ✨

**Comment by durkmurder** [comment]
bors merge

**Comment by durkmurder** [comment]
bors merge

**Comment by durkmurder** [comment]
bors retry

**Comment by durkmurder** [comment]
bors retry

**Comment by durkmurder** [comment]
bors retry

**Comment by durkmurder** [comment]
bors retry

**Comment by durkmurder** [comment]
bors retry

**Comment by durkmurder** [comment]
bors merge

**Comment by durkmurder** [comment]
bors retry

**Comment by durkmurder** [comment]
bors retry

**Comment by durkmurder** [comment]
bors merge

**Comment by jordanschalm** [comment]
@durkmurder Did you find some logs saying the SNJoinAndLeave test failed? When I look at the two failures, they are both timeouts and don't have any logs from running the tests I can see: 
```
Final attempt failed. Timeout of 900000ms hit
```

I am not sure where that 900000ms == 15 minute timeout is coming from. Before we quarantine the test could we look into the source of the timeout?

**Comment by jordanschalm** [comment]
It seems like it is from the `retry` tool:
```
Run nick-fields/retry@v2
  with:
    timeout_minutes: 15
    max_attempts: [2](https://github.com/onflow/flow-go/actions/runs/4325340436/jobs/7551368859#step:7:2)
```

**Comment by durkmurder** [comment]
In this job: https://github.com/onflow/flow-go/actions/runs/4325051958/jobs/7550691799, if you search by "FAIL:" you will find that it reports `TestEpochJoinAndLeaveSN`.

**Comment by jordanschalm** [comment]
The failure is caused by this:
```
suite.go:77: test panicked: Fail in goroutine after TestEpochJoinAndLeaveAN has completed
            goroutine 1251 [running]:
            runtime/debug.Stack()
            	/opt/hostedtoolcache/go/1.19.6/x64/src/runtime/debug/stack.go:24 +0x65
            github.com/stretchr/testify/suite.failOnPanic(0xc000b3c1a0, {0x1dd2700, 0xc002267440})
            	/home/runner/go/pkg/mod/github.com/stretchr/testify@v1.8.1/suite/suite.go:77 +0x3b
            github.com/stretchr/testify/suite.Run.func1.1()
            	/home/runner/go/pkg/mod/github.com/stretchr/testify@v1.8.1/suite/suite.go:161 +0x252
            panic({0x1dd2700, 0xc002267440})
            	/opt/hostedtoolcache/go/1.19.6/x64/src/runtime/panic.go:884 +0x212
            testing.(*common).Fail(0xc000d48680)
            	/opt/hostedtoolcache/go/1.19.6/x64/src/testing/testing.go:824 +0xe5
            testing.(*common).Fail(0xc000d48b60)
            	/opt/hostedtoolcache/go/1.19.6/x64/src/testing/testing.go:818 +0x47
            testing.(*common).FailNow(0xc000d48b60)
            	/opt/hostedtoolcache/go/1.19.6/x64/src/testing/testing.go:847 +0x2c
            testing.(*common).Fatalf(0xc000d48b60, {0x23ea412?, 0xc00439e998?}, {0xc0022673f0?, 0x3?, 0xc00439e901?})
            	/opt/hostedtoolcache/go/1.19.6/x64/src/testing/testing.go:955 +0x65
            github.com/dapperlabs/testingdock.(*Container).executeHealthCheck(0xc0008b9d90, {0x2973090?, 0xc000bc66c0?})
            	/home/runner/go/pkg/mod/github.com/dapperlabs/testingdock@v0.4.4/container.go:434 +0x27a
            github.com/dapperlabs/testingdock.(*Container).Start(0xc0008b9d90, {0x2973090?, 0xc000bc66c0})
            	/home/runner/go/pkg/mod/github.com/dapperlabs/testingdock@v0.4.4/container.go:269 +0xb78
            github.com/onflow/flow-go/integration/tests/epochs.(*Suite).runTestEpochJoinAndLeave(0xc00280fb00, 0x2, 0xc00439f6a8)
            	/home/runner/work/flow-go/flow-go/integration/tests/epochs/suite.go:677 +0x873
            github.com/onflow/flow-go/integration/tests/epochs.(*EpochJoinAndLeaveSNSuite).TestEpochJoinAndLeaveSN(0x0?)
            	/home/runner/work/flow-go/flow-go/integration/tests/epochs/epoch_join_and_leave_sn_test.go:22 +0x3c
            reflect.Value.call({0xc003089140?, 0xc0025191c8?, 0x22900dc?}, {0x225d246, 0x4}, {0xc001c8de70, 0x1, 0xc001c8dd80?})
            	/opt/hostedtoolcache/go/1.19.6/x64/src/reflect/value.go:584 +0x8c5
            reflect.Value.Call({0xc003089140?, 0xc0025191c8?, 0xc00280fb00?}, {0xc001c8de70?, 0x7f4f08a9c6f0?, 0xd0?})
            	/opt/hostedtoolcache/go/1.19.6/x64/src/reflect/value.go:368 +0xbc
            github.com/stretchr/testify/suite.Run.func1(0xc000b3c1a0)
            	/home/runner/go/pkg/mod/github.com/stretchr/testify@v1.8.1/suite/suite.go:175 +0x4b6
            testing.tRunner(0xc000b3c1a0, 0xc002a58990)
            	/opt/hostedtoolcache/go/1.19.6/x64/src/testing/testing.go:1446 +0x10b
            created by testing.(*T).Run
            	/opt/hostedtoolcache/go/1.19.6/x64/src/testing/testing.go:1493 +0x35f
```

The `testingdock` library panics after an earlier test successfully completes. I don't think it is related to `TestEpochJoinAndLeaveSN`.

Both `TestEpochJoinAndLeaveAN` and `TestEpochJoinAndLeaveSN` have 100% pass rate over the past 7 days:
https://dapperlabs.grafana.net/d/CE1MxVF7k/test-metrics-overview?orgId=1&var-latest_commit_sha=031ed1a9d64dee925e58a5e69e371fa0e2a913fb&var-latest_commit_date=2023-03-01%2017:46:33%2B00&var-dataset_name=dapperlabs-data.production_src_flow_test_metrics&var-skipped_tests_table=skipped_tests&var-test_results_table=test_results&var-query_limit=100000&var-test_category=integration-epochs&var-null_replacement=NULL

I would suggest we start by changing the lower timeout value [here](https://github.com/onflow/flow-go/blob/3cda6d2a6c40ffcca775e5d8c238ca562f7c691a/.github/workflows/ci.yml#L236) to 25 (matching the others).

**Comment by jordanschalm** [comment]
The error is the same as the epoch test, from `testingdock`. 

The failing test is: `--- FAIL: TestVerifySystemChunk (21.82s)`

But the actual error is: 
```
suite.go:77: test panicked: Fail in goroutine after TestHappyPath has completed
        goroutine 348 [running]:
        runtime/debug.Stack()
        	/opt/hostedtoolcache/go/1.19.6/x64/src/runtime/debug/stack.go:24 +0x65
        github.com/stretchr/testify/suite.failOnPanic(0xc002bd4000, {0x1c4f0e0, 0xc003ab0640})
        	/home/runner/go/pkg/mod/github.com/stretchr/testify@v1.8.1/suite/suite.go:77 +0x3b
        github.com/stretchr/testify/suite.recoverAndFailOnPanic(0xc000887040?)
        	/home/runner/go/pkg/mod/github.com/stretchr/testify@v1.8.1/suite/suite.go:72 +0x36
        panic({0x1c4f0e0, 0xc003ab0640})
        	/opt/hostedtoolcache/go/1.19.6/x64/src/runtime/panic.go:884 +0x212
        testing.(*common).Fail(0xc000887040)
        	/opt/hostedtoolcache/go/1.19.6/x64/src/testing/testing.go:824 +0xe5
        testing.(*common).FailNow(0xc000887040)
        	/opt/hostedtoolcache/go/1.19.6/x64/src/testing/testing.go:847 +0x2c
        testing.(*common).Fatalf(0xc000887040, {0x2257cad?, 0xc0026438c0?}, {0xc003ab05f0?, 0x40?, 0x0?})
        	/opt/hostedtoolcache/go/1.19.6/x64/src/testing/testing.go:955 +0x65
        github.com/dapperlabs/testingdock.(*Network).initialCleanup(0xc0024b1b80, {0x2786ab0, 0xc0026438c0})
        	/home/runner/go/pkg/mod/github.com/dapperlabs/testingdock@v0.4.4/network.go:137 +0x4d8
        github.com/dapperlabs/testingdock.(*Network).start(0xc0024b1b80, {0x2786ab0?, 0xc0026438c0})
        	/home/runner/go/pkg/mod/github.com/dapperlabs/testingdock@v0.4.4/network.go:46 +0x4e
        github.com/dapperlabs/testingdock.(*Suite).Start(0xc002bd4000?, {0x2786ab0?, 0xc0026438c0?})
        	/home/runner/go/pkg/mod/github.com/dapperlabs/testingdock@v0.4.4/suite.go:152 +0x1c5
        github.com/onflow/flow-go/integration/testnet.(*FlowNetwork).Start(0xc003e6b1e0, {0x2786ab0, 0xc0026438c0})
        	/home/runner/work/flow-go/flow-go/integration/testnet/network.go:244 +0x399
        github.com/onflow/flow-go/integration/tests/verification.(*Suite).SetupSuite(0xc003dd8600)
        	/home/runner/work/flow-go/flow-go/integration/tests/verification/suite.go:148 +0x1345
        github.com/stretchr/testify/suite.Run(0xc002bd4000, {0x2779658, 0xc003dd8600})
        	/home/runner/go/pkg/mod/github.com/stretchr/testify@v1.8.1/suite/suite.go:132 +0x39b
        github.com/onflow/flow-go/integration/tests/verification.TestVerifySystemChunk(0xc000a60000?)
        	/home/runner/work/flow-go/flow-go/integration/tests/verification/verify_system_chunk_test.go:11 +0x39
        testing.tRunner(0xc002bd4000, 0x24c5060)
        	/opt/hostedtoolcache/go/1.19.6/x64/src/testing/testing.go:1446 +0x10b
        created by testing.(*T).Run
        	/opt/hostedtoolcache/go/1.19.6/x64/src/testing/testing.go:1493 +0x35f
```

**Comment by durkmurder** [comment]
bors retry

**Pull request merged**